clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -212,25 +214,39 @@ public RecordAppendResult append(TopicPartition tp,
      * Abort the batches that have been sitting in RecordAccumulator for more than the configured requestTimeout
      * due to metadata being unavailable
      */
-    public List<RecordBatch> abortExpiredBatches(int requestTimeout, Cluster cluster, long now) {
+    public List<RecordBatch> abortExpiredBatches(int requestTimeout, Metadata metadata, long now) {
         List<RecordBatch> expiredBatches = new ArrayList<RecordBatch>();
         int count = 0;
         for (Map.Entry<TopicPartition, Deque<RecordBatch>> entry : this.batches.entrySet()) {
             Deque<RecordBatch> dq = entry.getValue();
-            synchronized (dq) {
-                // iterate over the batches and expire them if they have stayed in accumulator for more than requestTimeOut
-                Iterator<RecordBatch> batchIterator = dq.iterator();
-                while (batchIterator.hasNext()) {
-                    RecordBatch batch = batchIterator.next();
-                    // check if the batch is expired
-                    if (batch.maybeExpire(requestTimeout, now, this.lingerMs)) {
-                        expiredBatches.add(batch);
-                        count++;
-                        batchIterator.remove();
-                        deallocate(batch);
-                    } else {
-                        if (!batch.inRetry()) {
-                            break;
+            // Only expire a batch if its metadata is not available.
+            Cluster cluster = metadata.fetch();
+            // Check if the metadata might be old.
+            boolean maybeStaleMetadata = metadata.lastRefresh() > metadata.lastSuccessfulUpdate();
```

Comment:
@junrao Expiring all batches with RequestTimeout * Retries would help. But it seems not completely solve the probelem, right? For example, if the buffer is pretty full, it is still possible that the batches in the accumulator got timed out while some earlier batches are in flight. This might happen if the earlier batches got lucky and were sent just before they are timed out.

I am not sure if any other users are using Kafka for mission critical messaging. But after we started to use Kafka for DB replication (strict order, no data loss, low end to end latency), it seems that request timeout is not as rare as we thought before. If a batch is timed out before drained, the exception will be passed to the callback and the users have no choice but closing the producer. This has caused some trouble for us. So if possible, I would prefer to have this fixed in 0.10.0. For us it is actually OK because we are running off trunk and can pick it from the trunk as soon as the patch is checked in, but for people who are running with official Kafka release and interested in using Kafka for strict order messaging, having this issue might be a little disturbing.